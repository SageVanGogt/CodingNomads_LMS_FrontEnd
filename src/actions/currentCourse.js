export const updateCurrentCourse = course => ({
  type: 'UPDATE_CURRENT_COURSE',
  currentCourse: course
});
