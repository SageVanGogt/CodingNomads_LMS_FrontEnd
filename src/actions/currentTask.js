export const updateCurrentTask = (task) => ({
  type: 'UPDATE_CURRENT_TASK',
  currentTask: task
});
