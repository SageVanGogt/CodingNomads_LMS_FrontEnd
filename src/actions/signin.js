
export const signIn = (user) => {
  return ({
    type: 'SIGNIN_USER',
    user
  });
};