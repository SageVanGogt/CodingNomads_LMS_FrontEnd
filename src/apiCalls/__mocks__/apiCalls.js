import { mockDocs } from './../../mockData/mockDocs';
import { mockLabs } from './../../mockData/mockLabs';
import { mockTasks } from './../../mockData/mockTasks';
import { mockCourses } from './../../mockData/mockCourses';

export const getAllTasks = jest.fn().mockImplementation(
  () => Promise.resolve(mockTasks)
);

export const getTask = jest.fn();

export const postTask = jest.fn();

export const updateTask = jest.fn();

export const deleteTask = jest.fn();

export const getAllCourses = jest.fn().mockImplementation(
  () => Promise.resolve(mockCourses)
);

export const getCourse = jest.fn();

export const addCourse = jest.fn();

export const updateCourse = jest.fn();

export const deleteCourse = jest.fn();

export const getAllCoursesStudents = jest.fn();

export const addCourseStudent = jest.fn();

export const deleteCourseStudent = jest.fn();

export const getCourseTeachers = jest.fn();

export const addTeacherToCourse = jest.fn();

export const deleteTeacherFromCourse = jest.fn();

export const getCourseTasks = jest.fn();

export const postTaskToCourse = jest.fn();

export const deleteTaskFromCourse = jest.fn();

export const getAllDocs = jest.fn().mockImplementation(
  () => Promise.resolve(mockDocs)
);

export const getAllLabs = jest.fn().mockImplementation(
  () => Promise.resolve(mockLabs)
);

export const deleteLabsFromTask = jest.fn(); 

export const deleteDocsFromTask = jest.fn();

export const deleteTasksFromCourse = jest.fn();

export const deleteStudentsFromCourse = jest.fn();



