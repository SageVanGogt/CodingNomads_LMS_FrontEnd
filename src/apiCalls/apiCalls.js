import { apiPath } from '../constants/apiPath';

export const getAllTasks = async () => {
  const url = apiPath + '/tasks';
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error('Could not fetch tasks');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

// export const getAllTasksAuthed = async () => {
//   const token = localStorage.getItem('access_token') || null;
//   const url = apiPath + '/tasks';
//   const options = {
//     headers: { 'Authorization': `Bearer ${token}` }
//   }
//   try {
//     const response = await fetch(url);
//     if (response.status !== 200) {
//       throw Error('Could not fetch tasks');
//     } else {
//       return await response.json();
//     }
//   } catch (error) {
//     throw error;
//   }
// };

export const getTask = async (taskId) => {
  const url = apiPath + `/tasks/${taskId}`;
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error('Task could not be found.');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const postTask = async (newTask) => {
  const url = apiPath + '/tasks';
  const options = {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(newTask)
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 201) {
      throw Error('Could not post task');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
}; 

export const updateTask = async (updatedTask) => {
  const url = apiPath + `/tasks/${updatedTask.id}`;
  const options = {
    method: 'PATCH',
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(updatedTask)
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 200) {
      throw Error('Could not patch request');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const deleteTask = async (taskId) => {
  const url = apiPath + `/tasks/${taskId}`;
  const options = {
    method: 'DELETE',
    headers: { 'content-type': 'application/json' }
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 204) {
      throw Error('That id could not be found.');
    }
  } catch (error) {
    throw error;
  }
};

export const getAllCourses = async () => {
  const url = apiPath + '/courses';
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error('Could not get courses');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const getCourse = async (courseId) => {
  const url = apiPath + `/courses/${courseId}`;
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error('Could not find course');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const addCourse = async (newCourse) => {
  const url = apiPath + '/courses';
  const options = {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(newCourse)
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 200) {
      throw Error('Could not add course');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
}; 

export const updateCourse = async (updatedCourse) => {
  const url = apiPath + `/courses`;
  const options = {
    method: 'PATCH',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(updatedCourse)
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 200) {
      throw Error('Failed to update course.');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const deleteCourse = async (courseId) => {
  const url = apiPath + `/courses/${courseId}`;
  const options = {
    method: 'DELETE',
    headers: { 'content-type': 'application/json' }
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 204) {
      throw Error('That id could not be found.');
    }
  } catch (error) {
    throw error;
  }
};

export const getAllCourseStudents = async (courseId) => {
  const url = apiPath + `/courses/${courseId}/students`;
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error('Could not get students in course');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const addCourseStudent = async (newCourseStudent, course) => {
  const studentCourseRelation = {
    studentId: newCourseStudent.id,
    courseId: course.id
  };
  const url = apiPath + '/courses/students';
  const options = {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(studentCourseRelation)
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 201) {
      throw Error('Could not add student to course');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
}; 

export const deleteCourseStudent = async (student, course) => {
  const url = apiPath + `/courses/${course.id}/student/${student.id}`;
  const options = {
    method: 'DELETE',
    headers: { 'content-type': 'application/json' }
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 204) {
      throw Error('That id could not be found.');
    }
  } catch (error) {
    throw error;
  }
};

export const getCourseTeachers = async (courseID) => {
  const url = apiPath + `/courses/${courseID}/teachers`;
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error('Could not get teachers');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const addTeacherToCourse = async (courseId, teacherID) => {
  const url = apiPath + `/courses/${courseId}/teachers`;
  const options = {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(teacherID)
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 201) {
      throw Error('Could not add teacher to course');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const deleteTeacherFromCourse = async (courseId, teacherId) => {
  const url = apiPath + `/courses/${courseId}/teachers/${teacherId}`;
  const options = {
    method: 'DELETE',
    headers: { 'content-type': 'application/json' }
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 204) {
      throw Error('That id could not be found.');
    }
  } catch (error) {
    throw error;
  }
};

export const getCourseTasks = async (courseID) => {
  const url = apiPath + `/courses/${courseID}/tasks`;
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error('Could not get tasks for that course');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const postTaskToCourse = async (courseId, taskID) => {
  const url = apiPath + `/courses/${courseId}/tasks`;
  const options = {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(taskID)
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 201) {
      throw Error('Could not add task to course');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const deleteTaskFromCourse = async (courseId, taskID) => {
  const url = apiPath + `/courses/${courseId}/tasks/${taskID}`;
  const options = {
    method: 'DELETE',
    headers: { 'content-type': 'application/json' }
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 204) {
      throw Error('That id could not be found.');
    }
  } catch (error) {
    throw error;
  }
};

export const getAllDocs = async () => {
  const url = apiPath + '/docs';
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error('');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const getAllLabs = async () => {
  const url = apiPath + '/labs';
  try {
    const response = await fetch(url);
    if (response.status !== 200) {
      throw Error('');
    } else {
      return await response.json();
    }
  } catch (error) {
    throw error;
  }
};

export const deleteDocsFromTask = async (taskId, docsArr) => {
  const url = apiPath + `/task/${taskId}/docs`;
  const options = {
    method: 'DELETE',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ docs: docsArr })
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 204) {
      throw Error('That id could not be found.');
    }
  } catch (error) {
    throw error;
  }
};

export const deleteLabsFromTask = async (taskId, labsArr) => {
  const url = apiPath + `/task/${taskId}/labs`;
  const options = {
    method: 'DELETE',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ labs: labsArr })
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 204) {
      throw Error('That id could not be found.');
    }
  } catch (error) {
    throw error;
  }
};

export const deleteTasksFromCourse = async (courseId, tasksArr) => {
  const url = apiPath + `/course/${courseId}/tasks`;
  const options = {
    method: 'DELETE',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ tasks: tasksArr })
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 204) {
      throw Error('That id could not be found.');
    }
  } catch (error) {
    throw error;
  }
};

export const deleteStudentsFromCourse = async (courseId, studentsArr) => {
  const url = apiPath + `/course/${courseId}/students`;
  const options = {
    method: 'DELETE',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify({ students: studentsArr })
  };
  try {
    const response = await fetch(url, options);
    if (response.status !== 204) {
      throw Error('That id could not be found.');
    }
  } catch (error) {
    throw error;
  }
};

export const authUser = async (user) => {
  const url = apiPath + '/users'; 
  const options = {
    method: 'POST',
    headers: { 'content-type': 'application/json' },
    body: JSON.stringify(user)
  };
  try {
    const response = await fetch(url, options);
    if (response.message !== 'success') {
      throw Error('User does not exist');
    } 
    const userAuth = await response.json();
    return userAuth;
  } catch (error) {
    throw error;
  }
};


