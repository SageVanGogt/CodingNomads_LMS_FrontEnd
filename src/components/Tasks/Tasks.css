.Task_container {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  align-items: space-around;
  margin-left: 200px;
  width: calc(100vw - 200px);
  padding-top: 100px;
  background-color: yellow; }

.Task_new {
  position: fixed;
  top: 20px;
  right: 20px;
  width: 110px;
  height: 40px;
  background-color: #0F94FB;
  border-radius: 5px;
  font-weight: 800;
  font-size: .9rem; }

.Task_cards {
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  width: 80%;
  background-color: blue;
  border-radius: 5px; }
