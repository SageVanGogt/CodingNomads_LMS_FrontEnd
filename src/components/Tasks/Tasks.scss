.Task_container {
  display: flex;
  flex-wrap: wrap;
  justify-content: space-around;
  align-items: space-around;
  margin-left: 200px;
  width: calc(100vw - 200px);
  padding-top: 100px;
  background-color: #efefef;
}

.Task_new {
  position: fixed;
  top: 20px;
  right: 20px;
  width: 110px;
  height: 40px;
  background-color: #4885a2;
  border-radius: 5px;
  font-weight: 800;
  font-size: .9rem;
  color: white;
}

.Task_cards {
  display: flex;
  justify-content: center;
  flex-wrap: wrap;
  width: 80%;
  background-color: #efefef;
  border-radius: 5px;
}
