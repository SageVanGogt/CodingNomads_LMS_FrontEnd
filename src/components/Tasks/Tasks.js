import React, { Component} from 'react';
import { connect } from 'react-redux';
import { withRouter } from 'react-router-dom';
import { func } from 'prop-types';
import './Tasks.css';
import TaskCard from '../TaskCard/TaskCard';
import { mockTasks } from '../../mockData/mockTasks';
import { TASKS_EDIT } from '../../constants/routes';
import { postTask } from '../../apiCalls/apiCalls';
import { updateCurrentTask } from '../../actions/currentTask'

export class Tasks extends Component {
  constructor(props) {
    super(props);
    this.state = {
      allTasks: []
    }
  }

  addTask = async () => {
    // const newTask = await postTask()
    const newTask = {
      id: 1,
      name: '',
      description: '',
      videoLink: '',
      docs: [],
      labs: []
    };

    this.props.updateCurrentTask(newTask);
    this.props.history.push(TASKS_EDIT)
  }

  render() {
    const tasks = mockTasks.map((task, index) => {
      return (
        <TaskCard 
          key={`task-card-${index}`}
          {...task} 
        />
      );
    });

    return (
      <div className="Task_container">
        <button onClick={ this.addTask }>Add New Task</button>
        {tasks}
      </div>
    );
  }
}

export const mapDispatchToProps = (dispatch) => ({
  updateCurrentTask: (currentTask) => dispatch(updateCurrentTask(currentTask))
})

Tasks.propTypes = {
  updateCurrentTask: func.isRequired
}

export default withRouter(connect(null, mapDispatchToProps)(Tasks));

