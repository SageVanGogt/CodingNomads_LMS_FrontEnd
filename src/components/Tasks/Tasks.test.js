import React from 'react';
import { shallow } from 'enzyme';
import * as API from '../../apiCalls/apiCalls';
import { Tasks, mapDispatchToProps } from './Tasks';
import { updateCurrentTask } from '../../actions/currentTask';
jest.mock('../../apiCalls/apiCalls');

describe('Tasks', () => {
  let mockProps; 
  let mockTask;
  let wrapper;
    
  beforeEach(() => {
    mockProps = {
      updateCurrentTask: jest.fn(),
      history: { push: jest.fn() }
    }
    mockTask = {
      id: 1,
      name: '',
      description: '',
      videoLink: '',
      docs: [],
      labs: []
    }
    wrapper = shallow(<Tasks {...mockProps} />);
  });

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('addTask', () => {
    beforeEach(() => {
      API.postTask = jest.fn().mockImplementation(() => Promise.resolve(mockTask));
    })

    it('calls postTask', async () => {
      
    })

    // it('calls this.props.updateCurrentTask with the correct arguments', async () => {
    //   const expected = mockTask;
    //   const wrapperInst = wrapper.instance();
      
    //   await wrapperInst.addTask();

    //   expect(wrapperInst.props.updateCurrentTask).toHaveBeenCalledWith(expected);
    // })

    it ('calls this.props.history.push with the correct arguments', async () => {
      const expected = '/tasks/edit';
      const wrapperInst = wrapper.instance();

      await wrapperInst.addTask();

      expect(wrapperInst.props.history.push).toHaveBeenCalledWith(expected);

    })
  })

  // describe('mapDispatchToProps', () => {
  //   it('calls dispatch with the correct arguments', () => {
  //     const dispatch = jest.fn();
  //     const expected = updateCurrentTask(mockTask); 
  //     const mappedProps = mapDispatchToProps(dispatch);

  //     mappedProps.updateCurrentTask(mockTask);

  //     expect(dispatch).toHaveBeenCalledWith(expected);
  //   })
  // })
});
