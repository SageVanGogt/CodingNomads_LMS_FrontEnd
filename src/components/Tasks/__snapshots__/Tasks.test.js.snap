// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks should match the snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Tasks
    history={
        Object {
            "push": [Function],
          }
    }
    updateCurrentTask={[Function]}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <img
        height="200"
        src="/loading1.gif"
        width="200"
/>,
      "className": "Task_container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "height": "200",
        "src": "/loading1.gif",
        "width": "200",
      },
      "ref": null,
      "rendered": null,
      "type": "img",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <img
          height="200"
          src="/loading1.gif"
          width="200"
/>,
        "className": "Task_container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "height": "200",
          "src": "/loading1.gif",
          "width": "200",
        },
        "ref": null,
        "rendered": null,
        "type": "img",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
